rfcs/2020-11-06-4909-remap-template-fields.md

```diff
@@ -0,0 +1,211 @@
+current method of templating fields (known throughout this RFC as the template
+syntax).
+
+## Motivation
+
+Using the Remap language rather than the template syntax provides much greater
+power in how these fields are defined.
+
+The advantages of using Remap rather than the existing template syntax are:
+
+- One familiar syntax and function reference for Vector.
+- Access to all of remap's functions for templating.
+
+However, we do still need to support the template syntax in order to maintain
+backward compatability.
+
+## Internal Proposal
+
+There are two issues that need resolving to allow Remap to be used.
+
+### Use distinct syntax to distinguish between template and Remap syntax
+
+We will specify that double parentheses (`{{..}}`) are used for template
+syntax (the current technique) and triple parentheses (`{{{...}}}`) are used
+for Remap syntax.
+
+This would provide a clear and unambiguous way to distinguish between the
+syntaxes rather than rely on a set of heuristics.
+
+The script will need to return a value that can resolve to a String. `to_string`
+will be called on the final result. `String`, `Integer`, `Bool`, `Timestamp`
+will be allowed results. The script will, as much as possible, resolve this at
+loadtime. However, as event field types can resolve any type there will be some
+limitations to how much validation can be done.
+
+### Removing fields used in the template
+
+A feature available with the template syntax is to be able to list all the
+fields used within the template.
+
+The Loki sink wants a list of the fields that were used in the
+template. It uses this list to remove these fields from the event sent to Loki.
+
+If you had this sink:
+
+```toml
+[sinks.loki]
+  type = "loki"
+  inputs = ["json"]
+  endpoint = "http://localhost:3100"
+  remove_label_fields = true
+  labels.key = "{{foo}}-{{buzz}}"
+```
+
+And this event was sent to it:
+
+```json
+{"foo": "bar", "buzz": "zab", "message1": "thing", "message2": "thong"}
+```
+
+The actual message sent to Loki would be:
+
+```json
+{"message1": "thing", "message2": "thong"}
+```
+
+With the label `key = bar-zab` attached to the message. The fields `foo` and
+`buzz` that were used in defining the label have been removed from the message.
+
+It is not going to be possible to do this with the Remap syntax as it is much
+more complicated and ambiguous to determine which fields are used to generate
+the templated string.
+
+### Add `remap.before` and `remap.after` fields
+
+To replicate this functionality we will add `remap.before` and `remap.after`
+fields to the configuration of the source. (It will be useful in time to add
+these fields to every component within Vector).
+
+The user will be able to specify any fields they want dropped from the event
+in the `remap.after` script.
+
+```toml
+[sinks.loki]
+  type = "loki"
+  inputs = ["json"]
+  endpoint = "http://localhost:3100"
+  remove_label_fields = true
```

Comment:
Should we just drop this config straight away?

Even if people still use the original `{{ ... }}` templating, they still can replace `remove_label_fields` with `remap.after`. This change would be a breaking change, but one we're going to make at some point regardless, so might as well get it out of the way.